# Tidying up our data - Part 1

# Learning objectives

- Manipulate deeply nested json data and transform them into structured data ready to be loaded into a Data Warehouse.

## Loading our data from S3

In [0]:
filepath = "s3://full-stack-bigdata-datasets/Big_Data/YOUTUBE/songs.json"

In [0]:
df = spark.read.format('json').load(filepath, multiline=True)

## Tidying up

---

We have multiple issues with our data.  **It does not look like "tidy data" at all.**  
First, we have rows within rows...
And second, most of the data resides in deeply nested structure within the column items...

We will fix the former, then handle the latter in the next notebook.

### 1. Fixing the rows
Remember the `.explode` method? [documentation](https://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.functions.explode).  
What `.explode(...)` does, it "Returns a new row for each element in the given array or map." We will use a lot of it here!

If you remember properly, that's exactly the kind of structures we have in the schema of our DataFrame for the `items` column.

1. Print out the schema of `df`

In [0]:
df.printSchema()

root
 |-- etag: string (nullable = true)
 |-- items: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- contentDetails: struct (nullable = true)
 |    |    |    |-- caption: string (nullable = true)
 |    |    |    |-- contentRating: struct (nullable = true)
 |    |    |    |    |-- ytRating: string (nullable = true)
 |    |    |    |-- definition: string (nullable = true)
 |    |    |    |-- dimension: string (nullable = true)
 |    |    |    |-- duration: string (nullable = true)
 |    |    |    |-- licensedContent: boolean (nullable = true)
 |    |    |    |-- projection: string (nullable = true)
 |    |    |    |-- regionRestriction: struct (nullable = true)
 |    |    |    |    |-- allowed: array (nullable = true)
 |    |    |    |    |    |-- element: string (containsNull = true)
 |    |    |    |    |-- blocked: array (nullable = true)
 |    |    |    |    |    |-- element: string (containsNull = true)
 |    |    |-- etag: string (nullable = t

2. Import the PySpark SQL functions following usual convention

In [0]:
from pyspark.sql import functions as F

3. Use `.explode(...)` on the `items` column and count the number of results

In [0]:
df.withColumn('items', F.explode('items')).select('items').count()



Out[18]: 3907

If you got 3907 rows, you've made it, congrats! :)  
We will use this as our new working DataFrame:
- just do the same thing, but this time save the exploded dataset into a variable named `items_df`
- don't forget to give a proper alias to your newly compute column: `items`
- at the end, as a sanity check, make sure we have the right amount of columns in our new DataFrame

4. Follow previous instructions

In [0]:
items_df = df.withColumn('items', F.explode('items'))
items_df.select('items').count()


Out[19]: 3907

We're making progress, we now have one row per result (e.g. song)!

But each song is a deeply nested structure... We will take care of this in the following notebook.

5. Show the first 5 rows of the exploded dataset

In [0]:
items_df.limit(5).toPandas()

/databricks/spark/python/pyspark/sql/pandas/conversion.py:122: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Unable to convert the field items. If this column is not necessary, you may consider dropping it or converting to primitive type before the conversion.
Direct cause: Nested StructType not supported in conversion to Arrow
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.
  warn(msg)


etag 
 items 
 kind 
 pageInfo 
 
 
 
 
 0 
 U0fncx_GV9jD5SKQr15LMvwuPcs 
 ((false, (None,), sd, 2d, PT3M33S, True, recta... 
 youtube#videoListResponse 
 (38, 38) 
 
 
 1 
 U0fncx_GV9jD5SKQr15LMvwuPcs 
 ((false, (None,), hd, 2d, PT7M46S, False, rect... 
 youtube#videoListResponse 
 (38, 38) 
 
 
 2 
 U0fncx_GV9jD5SKQr15LMvwuPcs 
 ((false, (None,), sd, 2d, PT3M7S, False, recta... 
 youtube#videoListResponse 
 (38, 38) 
 
 
 3 
 U0fncx_GV9jD5SKQr15LMvwuPcs 
 ((false, (None,), hd, 2d, PT3M43S, False, rect... 
 youtube#videoListResponse 
 (38, 38) 
 
 
 4 
 U0fncx_GV9jD5SKQr15LMvwuPcs 
 ((false, (None,), hd, 2d, PT5M, False, rectang... 
 youtube#videoListResponse 
 (38, 38)

In [0]:
items_df.select("items").limit(5).display()

items List(List(false, List(null), sd, 2d, PT3M33S, true, rectangular, null), SqP7uUVSol30dxvuScN6JUny6T4, t1l8Z6gLPzo, youtube#video, List(10, UCUERSOitwgUq_37kGslN96w, VOLO, null, null, Enregistré et mixé par Cyrille PELTIER au "Keen Studio" à Tours.
Merci à Cyrille PELTIER et au "Keen Studio" à Tours.
Toutes les dates de la tournée sur volo.fr
Le nouvel album "Sans rire" est sorti le 11 mars 2013.
Volo était à l'Olympia le 29 avril dernier et est en tournée dans toute la France.
Le clip de "Toujours à côté" est sur youtube sur http://www.youtube.com/watch?v=cWCH2dpyw1c : Le premier extrait de l'album est en playliste sur France Bleue, Europe 1, RFM, RFI, Radio Alouette.
"Sans rire", autre extrait de l'album, est en playlist sur Virgin Radio.

Sortie digitale de "Toujours à côté"
https://itunes.apple.com/fr/album/toujours-a-cote-single/id570616846

Volo est un duo de frères. Frédéric et Olivier VOLOVITCH. 
Avec un nom de famille qui leur vient d'un arrière grand-père Ukrainien, nés au milieu d'une fratrie de 4 enfants, ils grandissent dans la campagne tourangelle dans une famille de profs.

Frédo intègre l'école de la rue Blanche en tant que comédien en 1992, Olivier le rejoint sur Paris en 1997 où il étudie la philosophie à la Sorbonne.
Le duo se forme officiellement en 2002.
Entre les différentes tournées des « Wriggles », dont Frédo est un des membres fondateurs, et Olivier le régisseur général, les deux frangins font leurs premiers concerts. D'abord et surtout dans des bars où autres Caf'conc, jusqu' à la sortie de leur 1er disque en 2005 chez « Opéra Music ».
Depuis, et en tout, les deux frangins, ont sorti cinq albums. Trois studios, un live acoustique et un best of. 
Peu médiatisé « Volo » est aussi et surtout un groupe de scène qui peut compter sur un public fidèle et grandissant. Produits en live par « Blue Line Productions », on les a vus en tournée dans toute la France, à la Cigale, au Trianon, au Bataclan, à l'Olympia, aux Francofolies de la Rochelle etc...
Aujourd'hui, récemment signé chez « Play On », Volo revient avec un nouvel album réalisé par Jean-François Delort sur des arrangements de Florian Dubos et Jocelyn Moze. Ce nouveau disque, attendu dans les bacs début 2013, sera accompagné d'une tournée qui démarrera cet automne et passera par la Cigale le 23 novembre 2012., none, List(Enregistré et mixé par Cyrille PELTIER au "Keen Studio" à Tours.
Merci à Cyrille PELTIER et au "Keen Studio" à Tours.
Toutes les dates de la tournée sur volo.fr
Le nouvel album "Sans rire" est sorti le 11 mars 2013.
Volo était à l'Olympia le 29 avril dernier et est en tournée dans toute la France.
Le clip de "Toujours à côté" est sur youtube sur http://www.youtube.com/watch?v=cWCH2dpyw1c : Le premier extrait de l'album est en playliste sur France Bleue, Europe 1, RFM, RFI, Radio Alouette.
"Sans rire", autre extrait de l'album, est en playlist sur Virgin Radio.

Sortie digitale de "Toujours à côté"
https://itunes.apple.com/fr/album/toujours-a-cote-single/id570616846

Volo est un duo de frères. Frédéric et Olivier VOLOVITCH. 
Avec un nom de famille qui leur vient d'un arrière grand-père Ukrainien, nés au milieu d'une fratrie de 4 enfants, ils grandissent dans la campagne tourangelle dans une famille de profs.

Frédo intègre l'école de la rue Blanche en tant que comédien en 1992, Olivier le rejoint sur Paris en 1997 où il étudie la philosophie à la Sorbonne.
Le duo se forme officiellement en 2002.
Entre les différentes tournées des « Wriggles », dont Frédo est un des membres fondateurs, et Olivier le régisseur général, les deux frangins font leurs premiers concerts. D'abord et surtout dans des bars où autres Caf'conc, jusqu' à la sortie de leur 1er disque en 2005 chez « Opéra Music ».
Depuis, et en tout, les deux frangins, ont sorti cinq albums. Trois studios, un live acoustique et un best of. 
Peu médiatisé « Volo » est aussi et surtout un groupe de scène qui peut compter sur un public fidèle et grandissant. Produits en live par « B

In [0]:
items_df.select("items").limit(5).show()

+--------------------+
|               items|
+--------------------+
|{{false, {null}, ...|
|{{false, {null}, ...|
|{{false, {null}, ...|
|{{false, {null}, ...|
|{{false, {null}, ...|
+--------------------+



## Wrap-up

You learned how to use `.explode(...)` to split arrays values into their own rows! 🎉